src/OrchardCore.Modules/OrchardCore.Media/Indexing/MediaFieldIndexHandler.cs

```diff
@@ -0,0 +1,77 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using OrchardCore.Indexing;
+using OrchardCore.Media.Fields;
+using OrchardCore.Media.Settings;
+using UglyToad.PdfPig;
+
+namespace OrchardCore.Media.Indexing
+{
+    public class MediaFieldIndexHandler : ContentFieldIndexHandler<MediaField>
+    {
+        private readonly IMediaFileStore _mediaFileStore;
+        private readonly IEnumerable<IMediaFileTextProvider> _mediaFileTextProviders;
+
+        public MediaFieldIndexHandler(
+            IMediaFileStore mediaFileStore,
+            IEnumerable<IMediaFileTextProvider> mediaFileTextProviders)
+        {
+            _mediaFileStore = mediaFileStore;
+            _mediaFileTextProviders = mediaFileTextProviders;
+        }
+
+        public async override Task BuildIndexAsync(MediaField field, BuildFieldIndexContext context)
+        {
+            var options = context.Settings.ToOptions();
+            var settings = context.ContentPartFieldDefinition.GetSettings<MediaFieldSettings>();
+
+            if (field.Paths.Length > 0)
+            {
+                if (settings.AllowMediaText)
+                {
+                    foreach (var key in context.Keys)
+                    {
+                        foreach (var mediaText in field.MediaTexts)
+                        {
+                            context.DocumentIndex.Set(key + ".MediaText", mediaText, options);
+                        }
+                    }
+                }
+
+                if (_mediaFileTextProviders.Any())
+                {
+                    // It doesn't really makes sense to store file contents without analyzing them for search as well.
+                    var fileIndexingOptions = options | DocumentIndexOptions.Analyze;
+
+                    foreach (var path in field.Paths)
+                    {
+                        using var fileStream = await _mediaFileStore.GetFileStreamAsync(path);
+                        if (fileStream != null)
+                        {
+                            var fileTexts = _mediaFileTextProviders
+                                .Where(provider => provider.CanHandle(path))
+                                .Select(provider => provider.GetText(path, fileStream));
```

Comment:
Here multiple providers can produce their own versions of the given file's text. Let's suppose these are different. Do you think @Skrypt it would make sense to index all such versions? An alternative is to pick the last (i.e. lastly registered) matching provider and produce the text with that only.

Reply:
What do you think, @Skrypt?

Reply:
I assume it would produce multiple versions if we had 2 different providers that CanHandle a .pdf file for example. Then, here I think it would be better to index only one version (lastly registered). So basically, this would allow to override the providers to use a different PDF parsing engine.

Reply:
Thanks!